### PR TITLE
Use Replica Override to manually scale a component

### PR DESCRIPTION
**Pull request by Richard87**
# TODO:
- [x] Deprecate "Start" feature, replace with "Reset scale" feature
- [x] Read component status from k8s deployment (keep outdated status if image doesnt match radix deployment)
- [ ] Update Go.mod with real versions (radix-operator)
- [ ] Release at the same time as the Operator
- [x] Add better labels to deployments, replace `is-job-aux-object` with `radix-aux-component` and `radix-auth-component-type=job-sleep`, look at possible changes to pods

### CODE REVIEW

File: api/deployments/models/component_deployment.go [hunk 1]
```diff
@@ -85,6 +85,14 @@ type Component struct {
 	// required: false
 	ReplicaList []ReplicaSummary `json:"replicaList"`
 
+	// Set if manuall control of replicas is in place. null means automatic controll, 0 means stopped and >= 1 is manually scaled.
```
Comment (satr) [inline api/deployments/models/component_deployment.go:88]:
```suggestion
	// Set if manual control of replicas is in place. Not set means automatic control, 0 means stopped and >= 1 is manually scaled.
```

File: api/environments/component_handler.go [hunk 2]
```diff
@@ -25,30 +27,49 @@ const (
...
+	if !radixutils.ContainsString(validaStatusesToScaleComponent, componentStatus) {
+		return environmentModels.CannotScaleComponent(appName, envName, componentName, componentStatus)
+	}
+	return eh.patchRadixDeploymentWithReplicas(ctx, updater, &replicas)
+}
 
-	log.Ctx(ctx).Info().Msgf("Stopping component %s, %s", componentName, appName)
+// ResetScaledComponent Starts a component
+func (eh EnvironmentHandler) ResetScaledComponent(ctx context.Context, appName, envName, componentName string, ignoreComponentStatusError bool) error {
+	log.Ctx(ctx).Info().Msgf("Resetting manuall scaled component %s, %s", componentName, appName)
```
Comment (satr) [inline api/environments/component_handler.go:52]:
```suggestion
	log.Ctx(ctx).Info().Msgf("Resetting manually scaled component %s, %s", componentName, appName)
```

File: api/deployments/models/component_status.go [hunk 3]
```diff
@@ -31,15 +42,47 @@ func (p ComponentStatus) String() string {
...
+	// Check if component is scaling up or down
+	if replicasUnavailable > 0 || replicas < replicasReady {
+		return ComponentReconciling
 	}
 
-	return status
+	return ConsistentComponent
+}
+
+func isCopmonentRestarting(component radixv1.RadixCommonDeployComponent, rd *radixv1.RadixDeployment) bool {
```
Comment (satr) [inline api/deployments/models/component_status.go:73]:
```suggestion
func isComponentRestarting(component radixv1.RadixCommonDeployComponent, rd *radixv1.RadixDeployment) bool {
```

File: api/deployments/models/component_status.go [hunk 4]
```diff
@@ -31,15 +42,47 @@ func (p ComponentStatus) String() string {
...
-	status := ConsistentComponent
+func ComponentStatusFromDeployment(component radixv1.RadixCommonDeployComponent, kd *appsv1.Deployment, rd *radixv1.RadixDeployment) ComponentStatus {
+	if kd == nil || kd.GetName() == "" {
+		return ComponentReconciling
+	}
+	replicasUnavailable := kd.Status.UnavailableReplicas
+	replicasReady := kd.Status.ReadyReplicas
+	replicas := pointers.Val(kd.Spec.Replicas)
+
+	if isCopmonentRestarting(component, rd) {
```
Comment (satr) [inline api/deployments/models/component_status.go:53]:
```suggestion
	if isComponentRestarting(component, rd) {
```

File: api/deployments/models/component_status.go [hunk 5]
```diff
@@ -31,15 +42,47 @@ func (p ComponentStatus) String() string {
...
+		return ComponentReconciling
 	}
 
-	return status
+	return ConsistentComponent
+}
+
+func isCopmonentRestarting(component radixv1.RadixCommonDeployComponent, rd *radixv1.RadixDeployment) bool {
+	restarted := component.GetEnvironmentVariables()[operatordefaults.RadixRestartEnvironmentVariable]
+	if strings.EqualFold(restarted, "") {
```
Comment (satr) [inline api/deployments/models/component_status.go:75]:
```suggestion
	if restarted == "" {
```
No need for case insesitive comparison

File: api/deployments/models/component_status.go [hunk 6]
```diff
@@ -31,15 +42,47 @@ func (p ComponentStatus) String() string {
...
+	restartedTime, err := commonutils.ParseTimestamp(restarted)
+	if err != nil {
+		log.Logger.Warn().Err(err).Msgf("unable to parse restarted time %v, component: %s", restarted, component.GetName())
+		return false
+	}
+	reconciledTime := rd.Status.Reconciled
+	if reconciledTime.IsZero() || restartedTime.After(reconciledTime.Time) {
+		return true
+	}
+	return false
```
Comment (satr) [inline api/deployments/models/component_status.go:87]:
```suggestion
	return reconciledTime.IsZero() || restartedTime.After(reconciledTime.Time)
```

**Review by satr — CHANGES_REQUESTED** [review]

**Review by satr — CHANGES_REQUESTED** [review]

File: api/environments/environment_handler.go [hunk 7]
```diff
@@ -319,26 +320,28 @@ func (eh EnvironmentHandler) StopEnvironment(ctx context.Context, appName, envNa
...
 	log.Ctx(ctx).Info().Msgf("Starting components in environment %s, %s", envName, appName)
 	for _, deployComponent := range radixDeployment.Spec.Components {
-		err := eh.StartComponent(ctx, appName, envName, deployComponent.GetName(), true)
-		if err != nil {
-			return err
+		if overide := deployComponent.GetReplicasOverride(); overide != nil && *overide == 0 {
+			err := eh.ResetScaledComponent(ctx, appName, envName, deployComponent.GetName(), true)
+			if err != nil {
+				return err
+			}
```
Comment (satr) [inline api/environments/environment_handler.go:336]:
```suggestion
		if override := deployComponent.GetReplicasOverride(); override != nil && *override == 0 {
			if err := eh.ResetScaledComponent(ctx, appName, envName, deployComponent.GetName(), true); err != nil {
				return err
			}
```

File: api/environments/environment_handler.go [hunk 8]
```diff
@@ -402,9 +405,11 @@ func (eh EnvironmentHandler) RestartApplication(ctx context.Context, appName str
 }
 
 func (eh EnvironmentHandler) getRadixCommonComponentUpdater(ctx context.Context, appName, envName, componentName string) (radixDeployCommonComponentUpdater, error) {
-	deploymentSummary, rd, err := eh.getRadixDeployment(ctx, appName, envName)
+	rd, err := kubequery.GetLatestRadixDeployment(ctx, eh.accounts.UserAccount.RadixClient, appName, envName)
 	if err != nil {
 		return nil, err
+	} else if rd == nil {
+		return nil, http.ValidationError(v1.KindRadixDeployment, "no radix deployments found")
 	}
```
Comment (satr) [inline api/environments/environment_handler.go:413]:
```suggestion
	}
	if rd == nil {
		return nil, http.ValidationError(v1.KindRadixDeployment, "no radix deployments found")
	}
```

File: api/kubequery/deployment.go [hunk 9]
```diff
@@ -19,3 +20,19 @@ func GetDeploymentsForEnvironment(ctx context.Context, client kubernetes.Interfa
 	}
 	return deployments.Items, nil
 }
+
+// GetDeploymentForComponent returns all the first deployment matching the specified application, environment and component name.
+func GetDeploymentForComponent(ctx context.Context, client kubernetes.Interface, appName, envName, componentName string) (*appsv1.Deployment, error) {
```
Comment (satr) [inline api/kubequery/deployment.go:25]: Is it in use?

File: api/environments/component_handler.go [hunk 10]
```diff
@@ -87,12 +108,12 @@ func (eh EnvironmentHandler) RestartComponent(ctx context.Context, appName, envN
 func (eh EnvironmentHandler) RestartComponentAuxiliaryResource(ctx context.Context, appName, envName, componentName, auxType string) error {
 	log.Ctx(ctx).Info().Msgf("Restarting auxiliary resource %s for component %s, %s", auxType, componentName, appName)
 
-	deploySummary, err := eh.deployHandler.GetLatestDeploymentForApplicationEnvironment(ctx, appName, envName)
+	radixDeployment, err := kubequery.GetLatestRadixDeployment(ctx, eh.accounts.UserAccount.RadixClient, appName, envName)
```
Comment (satr) [inline api/environments/component_handler.go:111]: Add docs comment

File: api/utils/predicate/kubernetes.go [hunk 11]
```diff
@@ -20,12 +20,21 @@ func IsAppAliasIngress(ingress networkingv1.Ingress) bool {
 }
 
 func IsPodForComponent(appName, componentName string) func(corev1.Pod) bool {
-	selector := labels.SelectorFromSet(radixlabels.Merge(radixlabels.ForApplicationName(appName), radixlabels.ForComponentName(componentName)))
+	selector := labels.SelectorFromSet(radixlabels.Merge(
+		radixlabels.ForApplicationName(appName),
+		radixlabels.ForComponentName(componentName)))
```
Comment (satr) [inline api/utils/predicate/kubernetes.go:25]:
```suggestion
	selector := labelselector.ForComponent(appName, componentName).AsSelector()
```

File: api/kubequery/deployment.go (hunk 9 above)
Comment (Richard87) [reply api/kubequery/deployment.go:25]: Thanks, removed :)

**Review by satr — DISMISSED** [review]

**Review by satr — DISMISSED** [review]

**Review by satr — APPROVED** [review]

**Review by satr — APPROVED** [review]

**Pull request merged**